src/cpp/rtps/builtin/discovery/endpoint/EDPSimple.cpp

```diff
@@ -147,48 +152,164 @@ bool EDPSimple::initEDP(
         logError(RTPS_EDP, "Problem creation SimpleEDP endpoints");
         return false;
     }
-#endif
+#endif // if HAVE_SECURITY
 
     return true;
 }
 
 //! Process the info recorded in the persistence database
-void EDPSimple::processPersistentData(t_p_StatefulReader & reader, t_p_StatefulWriter & writer)
+void EDPSimple::processPersistentData(
+        t_p_StatefulReader& reader,
+        t_p_StatefulWriter& writer,
+        key_list& demises)
 {
     std::lock_guard<RecursiveTimedMutex> guardR(reader.first->getMutex());
     std::lock_guard<RecursiveTimedMutex> guardW(writer.first->getMutex());
+    std::lock_guard<std::recursive_mutex> guardP(*mp_PDP->getMutex());
 
-    std::for_each(writer.second->changesBegin(),
-        writer.second->changesEnd(),
-        [&reader](CacheChange_t* change)
-    {
-        CacheChange_t* change_to_add = nullptr;
+    // own server instance
+    InstanceHandle_t server_key = mp_PDP->getLocalParticipantProxyData()->m_key;
 
-        if (!reader.first->reserveCache(&change_to_add, change->serializedPayload.length)) //Reserve a new cache from the corresponding cache pool
-        {
-            logError(RTPS_EDP, "Problem reserving CacheChange in EDPServer reader");
-            return;
-        }
+    // reference own references from writer history
+    std::forward_list<CacheChange_t*> removal;
+
+    // List known participants
+    key_list known_participants;
 
-        if (!change_to_add->copy(change))
+    std::for_each(
+        mp_PDP->ParticipantProxiesBegin(),
+        mp_PDP->ParticipantProxiesEnd(),
+        [&known_participants](const ParticipantProxyData* pD)
         {
-            logWarning(RTPS_EDP,"Problem copying CacheChange, received data is: " 
-                << change->serializedPayload.length << " bytes and max size in EDPServer reader"
-                << " is " << change_to_add->serializedPayload.max_size);
+            known_participants.insert(pD->m_key);
+        });
 
-            reader.first->releaseCache(change_to_add);
-            return ;
-        }
+    // We have not processed any PDP message yet but any lease duration callback may have modified demises
+    // already
```

Comment:
```suggestion
    // We have not processed any PDP message yet but any lease duration callback may have already modified 
    // demises
```

Reply:
Now it must be right: `already modified demises` 😄